Step-by-step-Tutorials/Custom-Storage-Providers.md

```diff
@@ -167,9 +168,9 @@ The format decisions have already been made, so coding up the `WriteStateAsync`
 
 
 ``` csharp
-public async Task WriteStateAsync(string grainType, Orleans.GrainReference grainRef, IGrainState grainState)
+public async Task WriteStateAsync(string grainType, GrainReference grainRef, IGrainState grainState)
 {
-    Dictionary<string, object> dataValues = grainState.AsDictionary();
+    IDictionary<string, object> dataValues = ((GrainState)grainState.State).AsDictionary();
     JavaScriptSerializer serializer = new JavaScriptSerializer();
```

Comment:
@jdom is correct here.  The IGrainState interface exposes an object via its State property.  The provider should store this object.  Providers should not cast the IGrainState to any concrete class.  A good example for this is the AzureBlobStorage storage provider.

The IGrainState also exposes string via its ETag property that should be used for concurrency.  If the etag of the grain state does not match that of the stored data, the write should fail.  If it matches, a new etag should be generated for the new data and updated in the IGrainState.  This must be atomic, so support for this must translate to the underlying storage.  If this can't be done using the underlying storage it should be noted in the provider documentation that concurrent data modification is not protected.